eval.go

```diff
@@ -450,34 +452,65 @@ func (d *indexData) List(ctx context.Context, q query.Q) (rl *RepoList, err erro
 		tr.Finish()
 	}()
 
+	var include func(rle *RepoListEntry) (bool, error)
+
 	q = d.simplify(q)
 	tr.LazyLog(q, true)
 	if c, ok := q.(*query.Const); ok {
-		return d.maybeRepoList(c.Value), nil
+		if !c.Value {
+			return &RepoList{}, nil
+		}
+		include = func(rle *RepoListEntry) (bool, error) {
+			return true, nil
+		}
+	} else {
+		// We need to run a search per repo to decide if it is included.
+		include = func(rle *RepoListEntry) (bool, error) {
+			qOneRepo := query.NewAnd(
+				query.NewRepoSet(rle.Repository.Name),
+				q)
+			sr, err := d.Search(ctx, qOneRepo, &SearchOptions{
+				ShardMaxMatchCount: 1,
+				TotalMaxMatchCount: 1,
+			})
+			if err != nil {
+				return false, err
+			}
+			return len(sr.Files) > 0, nil
+		}
 	}
 
-	sr, err := d.Search(ctx, q, &SearchOptions{
-		ShardMaxMatchCount: 1,
-		TotalMaxMatchCount: 1,
-	})
-	if err != nil {
-		return nil, err
+	var l RepoList
+
+	minimal := opts != nil && opts.Minimal
+	if minimal {
+		l.Minimal = make(map[uint32]*MinimalRepoListEntry, len(d.repoListEntry))
+	} else {
+		l.Repos = make([]*RepoListEntry, 0, len(d.repoListEntry))
 	}
 
-	return d.maybeRepoList(len(sr.Files) > 0), nil
-}
+	for i := range d.repoListEntry {
+		rle := &d.repoListEntry[i]
+		ok, err := include(rle)
+		if err != nil {
+			return nil, err
+		}
 
-// maybeRepoList returns a singleton repo list containing the repo for d if
-// include is true. Otherwise it returns an empty list.
-func (d *indexData) maybeRepoList(include bool) *RepoList {
-	l := &RepoList{}
-	if include {
-		l.Repos = make([]*RepoListEntry, 0, len(d.repoListEntry))
-		for _, rle := range d.repoListEntry {
-			l.Repos = append(l.Repos, &rle)
+		if !ok {
+			continue
+		}
+
+		if id := rle.Repository.ID(); id != 0 && minimal {
```

Comment:
it feels weird we need to parse the ID each time. I'm thinking we add the ID field and on unmarshalling repository from the shard we set the ID? Additionally we just start to set ID on Repository?

Reply:
Sounds good to me. Wasn't confident we wanted to do that, thanks for confirming!

Reply:
I'll do a follow up PR that sets the proper ID field in the repo metadata to limit the scope of this PR.

Reply:
I'll do a follow up PR that sets the proper ID field in the repo metadata to limit the scope of this PR.

Reply:
https://github.com/sourcegraph/zoekt/pull/99/commits/8e34ee2e879933487ad02d0a2ef03239f565ed3d

Reply:
https://github.com/sourcegraph/zoekt/pull/99/commits/8e34ee2e879933487ad02d0a2ef03239f565ed3d

Reply:
nice!